microcosm_flask/sentry.py

```diff
@@ -0,0 +1,67 @@
+from dataclasses import dataclass
+from os import environ
+from typing import Any, Optional
+from unittest.mock import MagicMock
+
+from microcosm.decorators import defaults
+from microcosm.object_graph import ObjectGraph
+from microcosm_logging.decorators import logger
+
+from sentry_sdk.utils import BadDsn
+
+
+@dataclass
+class SentryConfig:
+    dsn: Optional[str] = None
+    enabled: bool = False
+    client: Optional[Any] = None
+
+
+def before_send(event, hint):
+    """
+    Before sending a event to sentry scrub all values that are non-uuid data from the event
+    """
+    for exception in event["exception"]["values"]:
+        for frame in exception["stacktrace"]["frames"]:
+            for var_name, value in frame["vars"].items():
+                if var_name in ["self", "cls"]:
+                    continue
+                elif isinstance(value, str):
+                    if value.startswith("<function "):
+                        continue
+                    elif var_name.endswith("_id"):
+                        continue
+                frame["vars"][var_name] = "<redacted>"
+    return event
+
+
+@defaults(
+    enabled=False,
+    dsn=None,
+)
+@logger
+def configure_sentry(graph: ObjectGraph):
+    enabled, dsn, sentry = False, None, None
+    if graph.config.sentry_logging.enabled and graph.config.sentry_logging.dsn:
+        enabled = graph.config.sentry_logging.enabled
+        dsn = graph.config.sentry_logging.dsn
+        if graph.metadata.testing:
+            sentry = MagicMock()
+        else:
+            import sentry_sdk
+            try:
+                sentry = sentry_sdk.init(
+                    graph.config.sentry_logging.dsn,
+                    server_name=f"{graph.metadata.name}_api",
+                    release=graph.config.build_info_convention.sha1,
+                    environment=environ.get("MICROCOSM_ENVIRONMENT") or "undefined",
+                    before_send=before_send,
+                )
+            except BadDsn:
+                enabled, dsn, sentry = False, None, None
+                configure_sentry.logger.error("Bad DSN value set")
```

Comment:
what not to raise an exception?

Reply:
As this is configured from outside that app I didn't want an outside variable potentially disrupting/breaking the app loading. So opted to handle the case where bad config is supplied and disable the component completely.